pyat/at/tracking/particles.py

```diff
@@ -0,0 +1,131 @@
+    if not twiss:
+        cflag = False
+        if not argin.radiation:
+            argin.radiation_on()
+            cflag = True
+
+        emit0, beamdata, emit = ohmi_envelope(argin, refpts=[0])
+        sig_matrix = emit.r66[0]
+
+        if cflag:
+            argin.radiation_off()
+
+
+    else:
+        if len(argin) == 2:
+            [espread, blength] = argin
+            sig_matrix = numpy.array([
+                            [espread*espread, 0],
+                            [0, blength*blength]
+                                     ])
+
+        elif len(argin) == 3:
+            [bx, ax, epsx] = argin
+            sig_matrix = epsx * numpy.array([
+                                    [bx, -ax],
+                                    [-ax, (1 + ax * ax)/bx]
+                                            ])
+
+        elif len(argin) == 6:
+            [bx, ax, epsx, by, ay, epsy] = argin
+            sig_matrix = numpy.bmat([
+                            [sigma_matrix([bx, ax, epsx], twiss=True),
+                                numpy.zeros((2, 2))],
+                            [numpy.zeros((2, 2)),
+                                sigma_matrix([by, ay, epsy], twiss=True)]
+                                    ])
+
+        elif len(argin) == 8:
+            [bx, ax, epsx, by, ay, epsy, espread, blength] = argin
+            sig_matrix = numpy.bmat([
+                            [sigma_matrix([bx, ax, epsx], twiss=True),
+                                numpy.zeros((2, 4))],
+                            [numpy.zeros((2, 2)),
+                                sigma_matrix([by, ay, epsy], twiss=True),
+                                numpy.zeros((2, 2))],
+                            [numpy.zeros((2, 4)),
+                                sigma_matrix([espread, blength], twiss=True)]
+                                    ])
+
+    return sig_matrix
+
+
+def beam(np, sigma, orbit=None):
+    """
+    Generates an array of random particles according to the given sigma
+    matrix
+
+    PARAMETERS
+        np              Number of particles
+        sigma           sigma_matrix as calculated by at.sigma_matrix
+
+    KEYWORDS
+        orbit=None      An orbit can be provided to give a center of 
+                        mass offset to the distribution
+
+    OUTPUT
+        particle_dist   a matrix of shape (M, np) where M is shape of
+                        sigma matrix
+    """
+    v = numpy.random.normal(size=(sigma.shape[0], np))
+
+    try:
+        lmat = numpy.linalg.cholesky(sigma).T
+#        needs the .T to be the same as matlab
+
+    except numpy.linalg.LinAlgError:
+        row_idx = numpy.array([0, 1, 4, 5])
+        a1 = numpy.linalg.cholesky(sigma[row_idx[:, None], row_idx]).T
+        a = numpy.bmat([[a1, numpy.zeros((4, 2))], [numpy.zeros((2, 6))]])
+        row_idx = numpy.array([0, 1, 4, 5, 2, 3])
+        lmat = a[row_idx[:, None], row_idx]
+
+    particle_dist = numpy.squeeze(numpy.asarray(numpy.dot(lmat.T, v)))
```

Comment:
Same as above concerning `bmat`: if you use `block`, maybe you do not need the `asarray` call (to be checked)

Reply:
As above. Changing to numpy.block means I no longer need the asarray.